### DISCUSSION TITLE
How to get total documents count along with pagination

### DISCUSSION THREAD

**Discussion by aakashkag** [prisma/prisma#3087] [Q&A] [answered]
Hi,
Lets say we have user table.
Want to apply pagination on user table. 

```graphql
query{
  users(
    where:{
    Assigned_Role:{some:{user_role:{equals:"Student"}}}
    },
    skip:5
    take:5
    cursor:{user_id:5}
  ){
    user_id
    user_name
}
} 
```
Now I want to get total documents count along with pagination result. is it possible?

**Comment by ryands17** [comment, marked as answer]
Hey @aakashkag 👋 
You could perform this with a transaction here and return the result as per your schema:

```ts
await prisma.transaction([
    prisma.user.count(),
    prisma.user.findMany({
      take: 5,
      cursor: {
        id: 5,
      },
    }),
  ])
```

**Reply by ibadeeCodes** [reply]
It says :
_Invalid `prisma.post.aggregate()` invocation:\n\n\n  Failed to validate the query: `Field does not exist on enclosing type.` at `Query.aggregatepost.AggregatePost._count_

**Reply by mdperez86** [reply]
Then if I need to count how many items fit some filters I should perform 2 similar queries just to have the count. This is crazy!

```javascript
  await prisma.$transaction([
    prisma.product.count({
      where: {
        name: { contains: 'search' },
      },
    }),
    prisma.product.findMany({
      where: {
        name: { contains: 'search' },
      },
      orderBy: {
        name: "asc",
      },
      take: 200,
      skip: 10,
    }),
  ]);
```

Why not just have something like:

```javascript
  await prisma.product.findMany({
      where: {
        name: { contains: 'search' },
      },
      orderBy: {
        name: "asc",
      },
      _count: true,
      take: 200,
      skip: 10,
    });
```

For years paginated queries (as opposed to sliced queries) always return the total amount of items that fit certain criteria. Many ORMs did it that way because it's important for paginated systems who deals with complex sorting and filtering processes over billion of records.
I know I don't need sort the items to count them, but I need for sure the filters to do it as well as for pagination. So that means I will be executing 2 similar queries that could be heavy even when I should have only one.

**Reply by TristanWYL** [reply]
Agree with what @mdperez86 stated! The `count` field is quite common for pagination, which should better be retrieved within one query.

**Reply by TristanWYL** [reply]
Actually I have this need in my program. And for reducing the workload of the db, I achieved this within one query in this way:
```js
  let articles = await db.article.findMany(condition)
  const count = articles.length
  articles = articles.slice((page - 1) * PAGE_SIZE, page * PAGE_SIZE)
  return {
    articles,
    count,
  }
```

**Reply by ChristianRich** [reply]
> Actually I have this need in my program. And for reducing the workload of the db, I achieved this within one query in this way:
> 
> ```js
>   let articles = await db.article.findMany(condition)
>   const count = articles.length
>   articles = articles.slice((page - 1) * PAGE_SIZE, page * PAGE_SIZE)
>   return {
>     articles,
>     count,
>   }
> ```

This will not work for pagination queries using `skip` and `take` params since `articles.length` in your example will always equal or be less than the number of documents returned by the query.

We are still missing the total number of documents so we can calculate the total number of pages to display in the UI or make a pagination component.

[Example](https://miro.medium.com/max/638/1*F4X8aQyw3Ik7iINPg4AAbQ.png)

**Reply by TristanWYL** [reply]
> > Actually I have this need in my program. And for reducing the workload of the db, I achieved this within one query in this way:
> > ```js
> >   let articles = await db.article.findMany(condition)
> >   const count = articles.length
> >   articles = articles.slice((page - 1) * PAGE_SIZE, page * PAGE_SIZE)
> >   return {
> >     articles,
> >     count,
> >   }
> > ```
> 
> This will not work for pagination queries using `skip` and `take` params since `articles.length` in your example will always equal or be less than the number of documents returned by the query.
> 
> We are still missing the total number of documents so we can calculate the total number of pages to display in the UI or make a pagination component.
> 
> [Example](https://miro.medium.com/max/638/1*F4X8aQyw3Ik7iINPg4AAbQ.png)

Not really. You can easily convert `skip` and `take` to the `page` and `PAGE_SIZE` used in the code excerpt I posted. `articles.length` is always equal to the number of records queried under the condition. And the code will return only slice of the total queried records.

**Reply by ChristianRich** [reply]
My bad. I stand corrected.

I see what you're doing now. Instead of passing `skip` and `take` params to the query engine, you derive the paginated result set from the total result set, and then do your own pagination count.

I  like this approach, great little work-around.

Anybody foresee performance or memory issues implementing this approach?

**Reply by ionut-botizan** [reply]
So, rather than making 2 separate queries, you'd rather transfer the whole content of the table, with all the relations (potentially hundreds of megabytes or even gigabytes if you have *lots* of articles) over the network, load it into memory and then slice it! ...using Nodejs! ...for every single request coming from a user!
Yeah, that should work just fine! 😜

**Reply by ChristianRich** [reply]
> So, rather than making 2 separate queries, you'd rather transfer the whole content of the table, with all the relations (potentially hundreds of megabytes or even gigabytes if you have _lots_ of articles) over the network, load it into memory and then slice it! ...using Nodejs! ...for every single request coming from a user! Yeah, that should work just fine! 😜

That was my concern .... I'll stick with my two model query (one for the count and one for the actual query)

**Reply by ChristianRich** [reply]
Can we kindly request the authors of Prisma to address this? E.g include a `_count` param when doing paginated queries using `skip` and `take` params?

**Reply by zaverden** [reply]
issue exists: https://github.com/prisma/prisma/issues/7550

**Reply by transitive-bullshit** [reply]
I agree it's pretty annoying that this extremely common pattern requires two queries.

**Reply by motdde** [reply]
Thank you for sharing this. However, this method may not work when there are conditions involved. In such cases, you must ensure that the same condition is passed for the count to be accurate.

Edit: I am here wondering why we have take and skip options in count that only return a number. What will be the use case?

**Reply by benbatuu** [reply]
if you want to just a number of the data, you can just use : prisma.schemaName.count();

**Comment by ChristianRich** [comment]
Is there a solution to this problem that doesn't involve making two queries?

I'm aware that you can go `prisma.myModel.count()` but that's still an additional query and not efficient.
https://lifesaver.codes/answer/parameterized-count-queries-252

The only other solution I can think of is executing raw SQL but that kind of defy the purpose of using an ORM in the first place.

Pagination and total document count is a trivial DB use case and should be supported by the otherwise excellent Prisma project.

**Reply by cesarvspr** [reply]
please ping me if any update comes up

**Reply by andreasvh-conceto** [reply]
any ticket created for this issue? This would be awesome :)

**Comment by ymsaadi** [comment]
We need a better way to get count with pagination than the accepted answer.

**Comment by byoungd** [comment]
Any update for this? 
I think include `count` is so important. 

Hope team give us a better answer.

**Comment by agustints** [comment]
Any solution?

**Comment by omidh28** [comment]
Shouldn't this topic be moved to an issue?

**Reply by bsplosion** [reply]
Already was one, currently open: https://github.com/prisma/prisma/issues/7550

**Comment by laurent512** [comment]
up, requested feature for me as well :)

**Reply by vsinghipcium** [reply]
must have

**Comment by maple5233** [comment]
Is using `count` the only way currently?

**Comment by jaymeedwards** [comment]
This is really frustrating! Something like sequelize's findAndCountAll would be perfect here.

**Reply by jaymeedwards** [reply]
I may need to do an executeRaw for now. Which is kind of a bummer as it defeats the purpose of an ORM. 

If we had an override for findMany called findAndCountMany that returned your result set, plus added a column to the query “COUNT(*) AS PrismaResultCount” that was then returned as a count value in an associative array along with the full results as a rows value, that would seem like it may work.

I haven’t cracked open the code for how prisma generates queries yet so I’m guessing that’s an oversimplification.

**Reply by bsplosion** [reply]
@eavonius in case you're using that in production, note that you'll want to use a windowed a `count(*) over ()` instead of `count(*)` - the latter will _not_ return the total unpaginated count, just the count within the page (assuming you had aggregates).  The window function can be okay, though it have some significant performance implications for big datasets and can also have problems if the offset and/or limit are not defined to fit the data.

**Comment by edarcode** [comment]
yo emocionado con prisma y salen con esta bobada, no entiendo que les costaba contar la bendita consulta xD

**Comment by SC0d3r** [comment]
Would love to see this feature,  my suggestion is:
```javascript
const products = await prisma.product.findMany({
  skip: 15,
  take: 3,
  select: { /*...someSelections*/ }
  _count:{
    $total: true,
  }
})

const totalProducts = products._count.$total // this should give the total results
```

**Reply by Toshiro00** [reply]
or for each column. I don't know if there is already have feature like that.

```
_count: {
        $each: true
    }
```

**Comment by jmakinin** [comment]
this question is definitely not answered, 
the very answer marked as correct defeats the purpose of an orm making queries much more simplified and efficient. reason being that raw queries solve this issue rather better than prisma.
it should definitely be moved to an issue

**Reply by zaverden** [reply]
this is an issue https://github.com/prisma/prisma/issues/7550

**Comment by Zoot01** [comment]
Here is my suggested workaround as of now. Let me know if anyone has any better suggestions. 

```typescript
const getPaginatedResults = async (
  sortType: string | undefined | null,
  column: string,
  searchQuery: string,
  skip: number,
  take: number,
) => {
  const submission = await prisma.$transaction([
    prisma.submissions.count({
      where: {
        [column]: {
          mode: 'insensitive',
          contains: searchQuery,
        },
      },
    }),
    prisma.submissions.findMany({
      skip: skip,
      take: take,
      where: {
        [column]: {
          mode: 'insensitive',
          contains: searchQuery,
        },
      },
      orderBy: {
        [column]: sortType ?? 'asc',
      },
    }),
  ]);

  return {
    total: submission[0] ?? 0,
    data: submission[1],
  };
};
```

**Comment by mdinjamul** [comment]
This worked for me

```javascript
import { Prisma } from '@prisma/client';
import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient()

const findAll = async (userId: String) => {
  const query: Prisma.categoriesFindManyArgs = {
    where: {
      user_id: userId,
    }
  };
  const [categories, count] = await prisma.$transaction([
    prisma.categories.findMany(query),
    prisma.categories.count({ where: query.where })
  ]);

  return {
    pagination: {
      total: count
    },
    data: categories
  };
};
```

**Reply by kyle-turn2** [reply]
@injamulhaque535 how did you get this to work? I get a `Property 'UserFindManyArgs' does not exist on type 'typeof Prisma'.` I've tried importing `import { Prisma } from "@prisma/client";` and `import prisma from "@/lib/prisma";`. What does your `tsconfig.json` look like?

**Reply by kyle-turn2** [reply]
Ok, it does work, it's just that Typescript is not happy.

**Comment by maykon-oliveira** [comment]
I don't know if someone is still looking for a better solution for this, but I found a library in the Prisma documentation that gets rid of all this stuff.

[https://github.com/sandrewTx08/prisma-paginate](https://github.com/sandrewTx08/prisma-paginate)

**Reply by anhnch** [reply]
It calls 2 queries under the hood

**Comment by ndom91** [comment]
Based on [this](https://github.com/prisma/prisma/issues/7550#issuecomment-1614604789) comment in a similar issue in this repo, I wanted to share my Prisma client extension workaround for this issue,  and types are inferred correcltly as well :crossed_fingers: 


```ts
// /src/lib/db.ts

import { Prisma, PrismaClient } from '@prisma/client'

const prisma = new PrismaClient().$extends({
  name: 'findManyAndCount',
  model: {
    $allModels: {
      findManyAndCount<Model, Args>(
        this: Model,
        args: Prisma.Exact<Args, Prisma.Args<Model, 'findMany'>>
      ): Promise<[Prisma.Result<Model, Args, 'findMany'>, number]> {
        return prisma.$transaction([
          (this as any).findMany(args),
          (this as any).count({ where: (args as any).where })
        ]) as any;
      }
    }
  }
});

export default prisma
```


Usage:
```ts
// /src/components/bookmarks.tsx

const [data, count] = await prisma.bookmark.findManyAndCount({
  take: parseInt(limit) + parseInt(skip),
  skip: parseInt(skip),
  where: { userId: session?.user?.userId },
  include: {
    category: true,
    tags: { include: { tag: true } },
  },
  orderBy: { createdAt: "desc" },
})
```

**Comment by zwhitchcox** [comment]
@ndom91 The only problem with this is that it doesn't preserve the types for the regular prisma models. Solution:

```ts
import { Prisma, PrismaClient } from '../client/index.js';

const findManyAndCount = {
  name: 'findManyAndCount',
  model: {
    $allModels: {
      findManyAndCount<Model, Args>(
        this: Model,
        args: Prisma.Exact<Args, Prisma.Args<Model, 'findMany'>>
      ): Promise<
        [
          Prisma.Result<Model, Args, 'findMany'>,
          number,
          Args extends { take: number } ? number : undefined
        ]
      > {
        return prisma.$transaction([
          (this as any).findMany(args),
          (this as any).count({ where: (args as any).where }),
        ]) as any;
      },
    },
  },
};

function createExtendedClient() {
  return new PrismaClient(options)
    .$extends(findManyAndCount);
}

type ExtendedPrismaClient = ReturnType<typeof createExtendedClient>;
type PC = Omit<ExtendedPrismaClient, keyof PrismaClient> & PrismaClient

export const prisma =  createExtendedClient()  as unknown as PC;
```

The `Omit` is to prevent recursive max depth overflow (not sure why this happens though)

**Reply by ndom91** [reply]
Still works, nice!

**Reply by ndom91** [reply]
Did find a type error though now, it now claims my schema's don't have `findManyAndCount` on them :/

**Reply by zwhitchcox** [reply]
Ok (*panting*) in this code

* `findManyAndCount` and `findMany` types both work
* you can go to the definition of each type
* does not cause `maximum depth exceeded` errors when type checking

which I think fulfills all the requirements, at least the ones I have.

```ts
type FindManyAndCountType = typeof findManyAndCount.model.$allModels.findManyAndCount;

// Apply the inferred types to the extended models
type ModelsWithCustomMethods = {
  [Model in keyof PrismaClient]: PrismaClient[Model] extends { findMany: (...args: any[]) => Promise<any> }
    ? {
        findManyAndCount: FindManyAndCountType;
      } & PrismaClient[Model]
    : PrismaClient[Model];
};

type ExtendedPrismaClient = PrismaClient & ModelsWithCustomMethods;

function createExtendedClient(): ExtendedPrismaClient {
  const prisma = new PrismaClient(options) as unknown as ExtendedPrismaClient;
  prisma.$extends(findManyAndCount);
  return prisma;
}

const prisma: ExtendedPrismaClient = createExtendedClient();
```

to test the types:

```ts
// Testing the types
prisma.appointment.findManyAndCount({ where: { id: 1 } }); // Should work
prisma.appointment.findMany({ where: { id: 1 } }); // Should work
prisma.appointment.findManyAndCount({ whe: { id: 1 } }); // Should error
prisma.appointment.findMany({ we: { id: 1 } }); // Should error
```

@ndom91

**Reply by ndom91** [reply]
Ok (exhales) purrrfect :ok_hand: :joy:

**Reply by MkDierz** [reply]
hey im trying to use this with edge and prisma accelerate but dosent work
im using nextjs 14 with postgresql
```typescript
import { PrismaClient } from "@prisma/client/edge";
import { withAccelerate } from "@prisma/extension-accelerate";
import { Prisma } from "@prisma/client";

const findManyAndCount = {
  name: "findManyAndCount",
  model: {
    $allModels: {
      findManyAndCount<Model, Args>(
        this: Model,
        args: Prisma.Exact<Args, Prisma.Args<Model, "findMany">>
      ): Promise<
        [
          Prisma.Result<Model, Args, "findMany">,
          number,
          Args extends { take: number } ? number : undefined
        ]
      > {
        return prisma.$transaction([
          (this as any).findMany(args),
          (this as any).count({ where: (args as any).where }),
        ]) as any;
      },
    },
  },
};

type FindManyAndCountType =
  typeof findManyAndCount.model.$allModels.findManyAndCount;

type ModelsWithCustomMethods = {
  [Model in keyof PrismaClient]: PrismaClient[Model] extends {
    findMany: (...args: any[]) => Promise<any>;
  }
    ? {
        findManyAndCount: FindManyAndCountType;
      } & PrismaClient[Model]
    : PrismaClient[Model];
};

type ExtendedPrismaClient = PrismaClient & ModelsWithCustomMethods;

const prismaClientSingleton = () => {
  const prisma = new PrismaClient() as unknown as ExtendedPrismaClient;
  prisma.$extends(withAccelerate());
  prisma.$extends(findManyAndCount);
  return prisma;
};

declare global {
  var prismaGlobal: undefined | ReturnType<typeof prismaClientSingleton>;
}

const prisma = globalThis.prismaGlobal ?? prismaClientSingleton();

export default prisma;

if (process.env.NODE_ENV !== "production") globalThis.prismaGlobal = prisma;

```
```bash
 ⨯ src/lib/query.ts (5:35) @ findManyAndCount
 ⨯ TypeError: _db__WEBPACK_IMPORTED_MODULE_0__.default.wallet.findManyAndCount is not a function
    at readWallets (./src/lib/query.ts:9:75)
    at Page (wallet/page.tsx:15:79)
    at stringify (<anonymous>)
    at AsyncLocalStorage.run (node:async_hooks:346:14)
  3 |
  4 | export const readWallets = async (skip: number = 0, limit: number = 10) => {
> 5 |   const data = await prisma.wallet.findManyAndCount({
    |                                   ^
  6 |     take: limit + skip,
  7 |     skip: skip,
  8 |   });
```

**Reply by Grandnainconnu** [reply]
> hey im trying to use this with edge and prisma accelerate but dosent work im using nextjs 14 with postgresql
> 
> ```ts
> import { PrismaClient } from "@prisma/client/edge";
> import { withAccelerate } from "@prisma/extension-accelerate";
> import { Prisma } from "@prisma/client";
> 
> const findManyAndCount = {
>   name: "findManyAndCount",
>   model: {
>     $allModels: {
>       findManyAndCount<Model, Args>(
>         this: Model,
>         args: Prisma.Exact<Args, Prisma.Args<Model, "findMany">>
>       ): Promise<
>         [
>           Prisma.Result<Model, Args, "findMany">,
>           number,
>           Args extends { take: number } ? number : undefined
>         ]
>       > {
>         return prisma.$transaction([
>           (this as any).findMany(args),
>           (this as any).count({ where: (args as any).where }),
>         ]) as any;
>       },
>     },
>   },
> };
> 
> type FindManyAndCountType =
>   typeof findManyAndCount.model.$allModels.findManyAndCount;
> 
> type ModelsWithCustomMethods = {
>   [Model in keyof PrismaClient]: PrismaClient[Model] extends {
>     findMany: (...args: any[]) => Promise<any>;
>   }
>     ? {
>         findManyAndCount: FindManyAndCountType;
>       } & PrismaClient[Model]
>     : PrismaClient[Model];
> };
> 
> type ExtendedPrismaClient = PrismaClient & ModelsWithCustomMethods;
> 
> const prismaClientSingleton = () => {
>   const prisma = new PrismaClient() as unknown as ExtendedPrismaClient;
>   prisma.$extends(withAccelerate());
>   prisma.$extends(findManyAndCount);
>   return prisma;
> };
> 
> declare global {
>   var prismaGlobal: undefined | ReturnType<typeof prismaClientSingleton>;
> }
> 
> const prisma = globalThis.prismaGlobal ?? prismaClientSingleton();
> 
> export default prisma;
> 
> if (process.env.NODE_ENV !== "production") globalThis.prismaGlobal = prisma;
> ```
> 
> ```shell
>  ⨯ src/lib/query.ts (5:35) @ findManyAndCount
>  ⨯ TypeError: _db__WEBPACK_IMPORTED_MODULE_0__.default.wallet.findManyAndCount is not a function
>     at readWallets (./src/lib/query.ts:9:75)
>     at Page (wallet/page.tsx:15:79)
>     at stringify (<anonymous>)
>     at AsyncLocalStorage.run (node:async_hooks:346:14)
>   3 |
>   4 | export const readWallets = async (skip: number = 0, limit: number = 10) => {
> > 5 |   const data = await prisma.wallet.findManyAndCount({
>     |                                   ^
>   6 |     take: limit + skip,
>   7 |     skip: skip,
>   8 |   });
> ```

Got the same problem, solved it by doing:
```ts
function createExtendedClient(): ExtendedPrismaClient {
  const prisma = new PrismaClient({
    log:
      process.env.NODE_ENV === "development"
        ? ["query", "error", "warn"]
        : ["error"],
  }).$extends(findManyAndCount);

  return prisma as unknown as ExtendedPrismaClient;
}
```

By returning the `.$extends` result straight, it works.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
